tests/models/test_transformers.py

```diff
@@ -107,49 +124,58 @@ def test_phi3_transformers_orig():
     assert "5" in (output[0]["generated_text"])
 
 
-def test_phi3_loading(phi3_model: models.Model):
-    lm = phi3_model
-    lm += f"""You are a counting bot. Just keep counting numbers. 1,2,3,4, <|assistant|>"""
-    lm += gen("five", max_tokens=10)
-    assert "5" in lm["five"]
+# TODO: put this in the rest of the testing framework
 
 
-@pytest.mark.needs_credentials
-@pytest.mark.skip("Need to figure out auth")
-def test_llama3_chat():
-    lm = models.Transformers(
-        r"meta-llama/Meta-Llama-3-8B-Instruct", trust_remote_code=True
-    )
-    with system():
-        lm += "You are a counting bot. Just keep counting numbers."
+def test_phi3_chat_basic(phi3_model: models.Model):
+    lm = phi3_model
+
+    lm += "You are a counting bot. Just keep counting numbers."
     with user():
         lm += "1,2,3,4"
     with assistant():
         lm += gen(name="five", max_tokens=10)
 
     assert "5" in lm["five"]
 
+def test_phi3_chat_unrolled(phi3_model: models.Model):
+    lm = phi3_model
+    # Manually convert the chat format into completions style
+    lm += f"""<|user|>\nYou are a counting bot. Just keep counting numbers.<|end|><|assistant|>1,2,3,4,"""
```

Comment:
I think you need `\n<|assistant|>\n` here. Here's the template for reference:

{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') %}{{'<|user|>' + '
' + message['content'] + '<|end|>' + '
' + '<|assistant|>' + '
'}}{% elif (message['role'] == 'assistant') %}{{message['content'] + '<|end|>' + '
'}}{% endif %}{% endfor %}

Reply:
Phi-3 just has a really weird tokenizer. While the chat template inserts newlines, the tokenizer immediately strips them due to an rstrip they enabled on the special tokens. So functionally they are identical, and the rest of the tests with newlines/whitespaces everywhere are to check for this (and the instability it leads to). That said, this test should pass either way, but as-written is the closest thing to what gets passed to the model

Reply:
updated to be safe!